File: filly-core/src/widgets/password.rs
use crate::event::{Event, KeyCode};
use crate::render::{
    Alignment, BorderStyle, EdgeInsets, Rect, RenderNode, RenderTree, TextStyle,
};
use crate::widget::{EventResult, Widget};
use filly_protocol::WidgetResponse;

pub struct PasswordWidget {
    pub title: String,
    pub message: String,
    pub text: String,
    pub placeholder: String,
    dirty: bool,
}

impl PasswordWidget {
    pub fn new(title: String, message: String, placeholder: Option<String>) -> Self {
        Self {
            title,
            message,
            text: String::new(),
            placeholder: placeholder.unwrap_or_default(),
            dirty: true,
        }
    }
}

impl Widget for PasswordWidget {
    fn render(&self, area: Rect) -> RenderTree {
        let box_w = 60u16.min(area.width.saturating_sub(2));
        let box_h = 10u16.min(area.height.saturating_sub(2));
        let box_x = (area.width.saturating_sub(box_w)) / 2;
        let box_y = (area.height.saturating_sub(box_h)) / 2;

        let mut children = Vec::new();
        if !self.title.is_empty() {
            children.push(RenderNode::Text {
                rect: Rect::new(1, 0, box_w - 2, 1),
                content: self.title.clone(),
                alignment: Alignment::Center,
                style: TextStyle {
                    bold: true,
                    fg: Some("green".into()),
                    ..TextStyle::normal()
                },
                role: None,
                label: None,
                description: None,
            });
        }
        if !self.message.is_empty() {
            children.push(RenderNode::Text {
                rect: Rect::new(1, 1, box_w - 2, 2),
                content: self.message.clone(),
                alignment: Alignment::Left,
                style: TextStyle::normal(),
                role: None,
                label: None,
                description: None,
            });
        }

        let input_y = if self.message.is_empty() { 2 } else { 4 };
        children.push(RenderNode::Input {
            rect: Rect::new(1, input_y, box_w - 2, 1),
            text: self.text.clone(),
            cursor: 0,
            placeholder: self.placeholder.clone(),
            masked: true,
            role: None,
            label: None,
            description: None,
        });

        children.push(RenderNode::Text {
            rect: Rect::new(1, box_h - 1, box_w - 2, 1),
            content: "Type + Enter:confirm  Esc:cancel".into(),
            alignment: Alignment::Center,
            style: TextStyle::muted(),
            role: None,
            label: None,
            description: None,
        });

        RenderTree::Container {
            rect: Rect::new(box_x, box_y, box_w, box_h),
            background: None,
            border: BorderStyle::Single,
            padding: EdgeInsets::zero(),
            children,
            role: None,
            label: None,
            description: None,
        }
    }

    fn handle_event(&mut self, event: &Event) -> EventResult {
        match event {
            Event::Key(code, _) => match code {
                KeyCode::Esc => EventResult::Response(WidgetResponse {
                    result: None,
                    cancelled: true,
                    error: None,
                }),
                KeyCode::Enter => EventResult::Response(WidgetResponse {
                    result: Some(serde_json::Value::String(self.text.clone())),
                    cancelled: false,
                    error: None,
                }),
                KeyCode::Char(c) => {
                    self.text.push(*c);
                    self.dirty = true;
                    EventResult::Handled
                }
                KeyCode::Backspace => {
                    self.text.pop();
                    self.dirty = true;
                    EventResult::Handled
                }
                _ => EventResult::Unhandled,
            },
            _ => EventResult::Unhandled,
        }
    }

    fn is_dirty(&self) -> bool {
        self.dirty
    }
    fn clear_dirty(&mut self) {
        self.dirty = false;
    }
}

File: filly-core/widgets/password.c
#include <stdio.h>
#include "password.h"
#include <stdlib.h>
#include <string.h>

typedef struct {
    char *title;
    char *message;
    char *text;
    char *placeholder;
    bool dirty;
} PasswordData;

static void password_render(Widget *self, Rect area, RenderTree *out) {
    PasswordData *d = (PasswordData *)(self + 1);
    memset(out, 0, sizeof(*out));
    int box_w = 60, box_h = 10;
    if (box_w > area.w - 2) box_w = area.w - 2;
    if (box_h > area.h - 2) box_h = area.h - 2;
    int box_x = (area.w - box_w) / 2, box_y = (area.h - box_h) / 2;

    out->accessible.role = strdup("dialog");
    out->accessible.label = strdup(d->title ? d->title : "Password");

    RenderTree *children = calloc(3 + 1, sizeof(RenderTree));
    int idx = 0;
    if (d->title && strlen(d->title)) {
        RenderTree *t = &children[idx++];
        t->type = RNODE_TEXT;
        t->rect = rect_new(1, 0, box_w - 2, 1);
        t->text.content = strdup(d->title);
        t->text.align = ALIGN_CENTER;
        t->text.style = textstyle_selected();
        t->accessible.role = strdup("heading");
        t->accessible.label = strdup(d->title);
    }
    int input_y = 2;
    if (d->message && strlen(d->message)) {
        RenderTree *t = &children[idx++];
        t->type = RNODE_TEXT;
        t->rect = rect_new(1, 1, box_w - 2, 2);
        t->text.content = strdup(d->message);
        t->text.align = ALIGN_LEFT;
        t->text.style = textstyle_normal();
        t->accessible.role = strdup("description");
        input_y = 4;
    }
    RenderTree *inp = &children[idx++];
    inp->type = RNODE_INPUT;
    inp->rect = rect_new(1, input_y, box_w - 2, 1);
    inp->input.text = d->text ? strdup(d->text) : strdup("");
    inp->input.cursor = 0;
    inp->input.placeholder = d->placeholder ? strdup(d->placeholder) : strdup("");
    inp->input.masked = true;
    inp->accessible.role = strdup("textbox");
    inp->accessible.label = strdup("Password field");

    RenderTree *footer = &children[idx++];
    footer->type = RNODE_TEXT;
    footer->rect = rect_new(1, box_h - 2, box_w - 2, 1);
    footer->text.content = strdup("Type + Enter:confirm  Esc:cancel");
    footer->text.align = ALIGN_CENTER;
    footer->text.style = textstyle_muted();

    out->type = RNODE_CONTAINER;
    out->rect = rect_new(box_x, box_y, box_w, box_h);
    out->container.border = BORDER_SINGLE;
    out->container.padding = edgeinsets_zero();
    out->container.children = children;
    out->container.child_count = idx;
}

static EventResult password_handle_event(Widget *self, Event *ev, Backend *backend) {
    (void)backend;
    PasswordData *d = (PasswordData *)(self + 1);
    if (ev->type != EVENT_KEY) return event_result_unhandled();
    switch (ev->code) {
        case KEY_ESC:
            return event_result_response((WidgetResponse){ .result = NULL, .cancelled = true, .error = NULL });
        case KEY_ENTER:
            return event_result_response((WidgetResponse){ .result = cJSON_CreateString(d->text), .cancelled = false, .error = NULL });
        case KEY_CHAR: {
            int len = strlen(d->text);
            d->text = realloc(d->text, len + 2);
            d->text[len] = ev->ch;
            d->text[len + 1] = '\0';
            d->dirty = true;
            return event_result_handled();
        }
        case KEY_BACKSPACE:
            if (strlen(d->text) > 0) {
                d->text[strlen(d->text) - 1] = '\0';
                d->dirty = true;
            }
            return event_result_handled();
        default:
            return event_result_unhandled();
    }
}

static bool password_is_dirty(Widget *self) { return ((PasswordData *)(self + 1))->dirty; }
static void password_clear_dirty(Widget *self) { ((PasswordData *)(self + 1))->dirty = false; }
static void password_destroy(Widget *self) {
    PasswordData *d = (PasswordData *)(self + 1);
    free(d->title); free(d->message); free(d->text); free(d->placeholder);
}

Widget *password_widget_new(const char *title, const char *message, const char *placeholder) {
    Widget *w = calloc(1, sizeof(Widget) + sizeof(PasswordData));
    w->vtable.render = password_render;
    w->vtable.handle_event = password_handle_event;
    w->vtable.is_dirty = password_is_dirty;
    w->vtable.clear_dirty = password_clear_dirty;
    w->vtable.destroy = password_destroy;
    PasswordData *d = (PasswordData *)(w + 1);
    d->title = title ? strdup(title) : NULL;
    d->message = message ? strdup(message) : NULL;
    d->text = strdup("");
    d->placeholder = placeholder ? strdup(placeholder) : strdup("");
    d->dirty = true;
    return w;
}